block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockDagKeyValueStorage.scala

```diff
@@ -53,14 +53,71 @@ final class BlockDagKeyValueStorage[F[_]: Concurrent: Log] private (
     def children(blockHash: BlockHash): F[Option[Set[BlockHash]]] =
-      childMap.get(blockHash).pure[F]
+      childMap.get(blockHash).map(_ intersect dagSet).pure[F]
 
     def latestMessageHash(validator: Validator): F[Option[BlockHash]] =
       latestMessagesMap.get(validator).pure[F]
 
     def latestMessageHashes: F[Map[Validator, BlockHash]] = latestMessagesMap.pure[F]
 
-    def invalidBlocks: F[Set[BlockMetadata]] = invalidBlocksSet.pure[F]
+    def invalidBlocks: F[Set[BlockMetadata]] =
+      invalidBlocksSet.filter(m => dagSet.contains(m.blockHash)).pure[F]
+
+    override def truncate(
+        latestMessages: Map[Validator, BlockHash]
+    ): F[BlockDagRepresentation[F]] = {
+      final case class HeightView(
+          height: Long,
+          excess: Iterable[ByteString],
+          inView: Iterable[ByteString]
+      )
+      for {
+        lmMetas <- latestMessages.values.toStream
+                    .traverse(lookup)
+                    .map(_.map(_.get).toSet)
+        lmHeights            = lmMetas.map(_.blockNum)
+        highestHeight        = lmHeights.max
+        lowestHeight         = lmHeights.min
+        lmPerValidatorHeight = lmMetas.map(m => (m.sender -> m.blockNum)).toMap
+
+        // messages higher then highest latest message are outside the view
+        (highExcess, withoutHighExcess) = heightMap.partition { case (h, _) => h > highestHeight }
+        // messages inside range of latest message heights might contain new messages from validator that are not seen yet
+        lmRangeViews <- heightMap
+                         .range(lowestHeight, highestHeight)
+                         .toList
+                         .traverse {
+                           case (h, messages) =>
+                             for {
+                               lms <- messages.toList
+                                       .traverse(lookup)
+                                       .map(_.map(_.map(m => (m.blockHash, m.sender)).get))
+                               (excess, inView) = lms.partition {
+                                 case (_, sender) => h > lmPerValidatorHeight(sender)
+                               }
+                               r = HeightView(
+                                 h,
+                                 excess = excess.map { case (hash, _) => hash },
+                                 inView = inView.map { case (hash, _) => hash }
+                               )
+
+                             } yield r
+                         }
+        // messages below lowest height across input latest messages are always in view,
+        // as it was decided so the whole excess is this sum
+        excessMessages  = highExcess.values.flatten ++ lmRangeViews.flatMap(_.excess)
+        truncatedDagSet = this.dagSet diff excessMessages.toSet
+
+        truncatedHeightMap = lmRangeViews.foldLeft(withoutHighExcess)(
+          (acc, heightView) => acc.updated(heightView.height, heightView.inView.toSet)
```

Comment:
`heightView.inView` is an iterator and here it's iterated for each element in _lmRangeViews_. Think always how to minimize iteration over collections you are working with.

Reply:
@tgrospic It's `Iterable`, not `Iterator`. So this means that this is just a collection, right?

Reply:
Yes, you are right I think of Iterator but Iterable has only one method to provide Iterator.

Iterator is basically these two functions. This means the only way to do anything with this "collection" is to get next element (maybe all).
https://github.com/scala/scala/blob/2.12.x/src/library/scala/collection/Iterator.scala#L359-L380
```scala
trait Iterator[+A] extends TraversableOnce[A] {
  /** Tests whether this iterator can provide another element.
   *
   *  @return  `true` if a subsequent call to `next` will yield an element,
   *           `false` otherwise.
   *  @note    Reuse: $preservesIterator
   */
  def hasNext: Boolean

  /** Produces the next element of this iterator.
   *
   *  @return  the next element of this iterator, if `hasNext` is `true`,
   *           undefined behavior otherwise.
   *  @note    Reuse: $preservesIterator
   */
  def next(): A
```

So the question is how to choose the right type of collection or better to say what kind of interface should provide this collection. The main thing we want to consider is performance (do only necessary work) but also that code is easy to understand which will reduce bugs.
**For collections better performance === less iterations.** 

For example
`inView = inView.map { case (hash, _) => hash }`
when executed here has one iteration to extract `Iterable[ByteString]` so when Set is needed here
`(acc, heightView) => acc.updated(heightView.height, heightView.inView.toSet)`
it needs to iterate once again to create a Set. This is done with builder implementation which uses TraversableOnce created from iterator.
https://github.com/scala/scala/blob/2.12.x/src/library/scala/collection/immutable/Set.scala#L372-L384

So here, more precise type for `inView` would be a Set because this is what is needed at the end when is added to height map. So when inView is created it will be converted to Set.
But even this will still be the same. :) `map` will iterate and toSet will iterate.
```scala
// inView: Set[ByteString]
inView = inView.map { case (hash, _) => hash }.toSet
```
So to make it *lazy*, first we need to convert it to iterator which will be used by map and toSet for each element only once.
```scala
// inView: Set[ByteString]
inView = inView.toIterator.map { case (hash, _) => hash }.toSet
```
---

Another view can be, why not use iterator as a type for `inView` and have the same effect. So here mapping will not happen but only when toSet is called later. This is also correct but I'd say little harder to understand and follow.
```scala
// inView: Iterator[ByteString]
inView = inView.toIterator.map { case (hash, _) => hash }
```